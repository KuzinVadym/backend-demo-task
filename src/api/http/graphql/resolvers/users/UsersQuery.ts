import {
  Arg,
  Ctx,
  FieldResolver,
  Int,
  Query,
  Resolver,
  Root
} from 'type-graphql';
import { UserType } from './UserType';
import { withHttpService } from '../../../shared/utils/withHttpService';

@Resolver((of) => UserType)
export class UsersQuery {
  @Query(() => [UserType])
  public async users(@Ctx() ctx): Promise<UserType[]> {
    const userHttpService = withHttpService('users', ctx.getState);
    return userHttpService.getUsers();
  }

  @Query(() => UserType)
  public async user(
    @Arg('id', (type) => Int) id: number,
    @Ctx() ctx
  ): Promise<UserType> {
    const userHttpService = withHttpService('users', ctx.getState);
    return userHttpService.getUser(id);
  }

  @FieldResolver()
  async createdTasks(@Root() user: UserType, @Ctx() ctx) {
    const userHttpService = withHttpService('tasks', ctx.getState);
    return userHttpService.getTasks({ createdByUserId: user.id });
  }

  @FieldResolver()
  async updatedTasks(@Root() user: UserType, @Ctx() ctx) {
    const userHttpService = withHttpService('tasks', ctx.getState);
    return userHttpService.getTasks({ updatedByUserId: user.id });
  }

  @FieldResolver()
  async assignedTasks(@Root() user: UserType, @Ctx() ctx) {
    const userHttpService = withHttpService('tasks', ctx.getState);
    return userHttpService.getTasks({ assignedUserId: user.id });
  }
}
