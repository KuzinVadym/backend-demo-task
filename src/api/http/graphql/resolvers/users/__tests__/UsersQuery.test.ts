import 'reflect-metadata';
import {
  ApolloServerTestClient,
  createTestClient
} from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-express';
import { buildSchema } from 'type-graphql';
import { UsersQuery } from '../UsersQuery';

const contextMock = {
  getState: jest.fn()
};

const usersServiceMock = {
  getUsers: jest.fn(),
  getUser: jest.fn()
};

const tasksServiceMock = {
  getTasks: jest.fn()
};

describe('UsersQuery', () => {
  let server: ApolloServer;

  beforeAll(async () => {
    const schema = await buildSchema({ resolvers: [UsersQuery] });
    server = new ApolloServer({
      schema,
      context: contextMock
    });
  });

  beforeEach(async () => {
    usersServiceMock.getUsers.mockResolvedValue([
      { login: 'login', email: 'email' }
    ]);
    usersServiceMock.getUser.mockResolvedValue({
      id: 1,
      login: 'login',
      email: 'email'
    });
    tasksServiceMock.getTasks.mockResolvedValue([{ id: 1 }]);

    contextMock.getState.mockReturnValue({
      httpServices: {
        users: usersServiceMock,
        tasks: tasksServiceMock
      }
    });
  });

  it('returns users data', async () => {
    const testQuery = `{
      users {
        login
        email
      }
    }
  `;

    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: {} });
    expect(res).toMatchSnapshot();
  });

  it('returns user data', async () => {
    const testQuery = `
    query UserQuery($id: Int!){
      user(id: $id) {
        login
        email
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: { id: 1 } });
    expect(res).toMatchSnapshot();
  });

  it('cals service with right params', async () => {
    const testQuery = `
    query UserQuery($id: Int!){
      user(id: $id) {
        id
        login
        email
        createdTasks{
          id
        }
        updatedTasks{
          id
        }
        assignedTasks{
          id
        }
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    await query({ query: testQuery, variables: { id: 1 } });

    expect(usersServiceMock.getUser).toBeCalledWith(1);
    expect(tasksServiceMock.getTasks).toBeCalledWith({ createdByUserId: 1 });
    expect(tasksServiceMock.getTasks).toBeCalledWith({ updatedByUserId: 1 });
    expect(tasksServiceMock.getTasks).toBeCalledWith({ assignedUserId: 1 });
  });
});
