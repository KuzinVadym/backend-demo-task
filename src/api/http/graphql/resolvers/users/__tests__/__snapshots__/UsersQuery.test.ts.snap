// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersQuery returns user data 1`] = `
Object {
  "data": Object {
    "user": Object {
      "email": "email",
      "login": "login",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`UsersQuery returns users data 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "email": "email",
        "login": "login",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
